updated _write_args on PythonVirtualenvOperator

maganaluis opened the pull request:
This change will ensure that argument provide_context=True works on v1.10.10 for PythonVirtualenvOperator

This permits the user use the configuration that is being passed at runtime through the UI, CLI or REST API. 

```python 
def do_work(**kwargs):    
    table_name = kwargs['dag_run'].conf.get('parameters')    
    # Rest of the code
```

Related bug:
https://github.com/apache/airflow/issues/8177
---
Make sure to mark the boxes below before creating PR: [x]

- [x] Description above provides context of the change
- [x] Unit tests coverage for changes (not needed for documentation changes)
- [x] Commits follow "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)"
- [x] Relevant documentation is updated including usage instructions.
- [x] I will engage committers as explained in [Contribution Workflow Example](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#contribution-workflow-example).

---
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/master/UPDATING.md).
Read the [Pull Request Guidelines](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#pull-request-guidelines) for more information.

yuqian90 committed:
[AIRFLOW-5391] Do not re-run skipped tasks when they are cleared (#7276)

If a task is skipped by BranchPythonOperator, BaseBranchOperator or ShortCircuitOperator and the user then clears the skipped task later, it'll execute. This is probably not the right
behaviour.

This commit changes that so it will be skipped again. This can be ignored by running the task again with "Ignore Task Deps" override.

(cherry picked from commit 1cdab56a6192f69962506b7ff632c986c84eb10d)

jloehel commented on airflow/operators/python_operator.py:
```diff
@@ -330,13 +330,28 @@ def _write_string_args(self, filename):
+        else:
+            serializer = pickle
+        # some args from context can't be loaded in virtual env
+        invalid_args = set(['dag', 'task', 'ti'])
```
Why is it not possible to load `ti`?

mik-laj replied:
TI is a SQL Alchemy model. It is very likely that it contains references to another object through the`task` attribute.

jloehel replied:
Thanks mik-laj. This means that xcoms are still not working for PythonVirtualenvOperators, right? Can I inject the system site-packages into the virtualenv to get it working?

mik-laj replied:
This is slightly problematic. Technically this is possible, but you would have to create the TaskInstance object from scratch. This can be a simple code that will only retrieve the object using ``session.query(TaskInstance)``

jloehel replied:
I see, it's tricky. Thanks for the detailed answers. I'll try to retrieve the TaskInstance or Xcom directly as a workaround.

maganaluis replied:
@jloehel We have not encountered any cases where to use Xcom, so I'm not familiar with it. Let me know if you want me to make any modification to this code, or you'll be making changes on your side.

jloehel replied:
Maybe I am using it wrong but I don't get the idea why it is an anti-pattern to exchange data between tasks. I am using this feature really often. I will make the changes on my side but I would love to see your PR merged soon. Thanks a lot for your work :-)

maganaluis replied:
@jloehel I don't mind if the PR doesn't merge, as long as PythonVirtualenvOp gets fixed, I think that will help a lot people. Feel free to send me an e-mail if you would like to follow up in case I need to modify that function again. Should be in my profile. Cheers :)

alexsstock replied:
This helped me thanks @maganaluis

ashb replied:
XCom isn't likely to be a problem, as you don't pass XCom rows/objects around, but instead the _value stored inside that row_.

Fokko commented on airflow/operators/python_operator.py:
```diff
@@ -330,13 +330,28 @@ def _write_string_args(self, filename):
+            for key, value in self.op_kwargs.items():
+                try:
+                    serializer.loads(serializer.dumps(value))
+                    if key not in invalid_args:
```
Why skip the invalid arguments here? It would make more sense to me to not serialize the set above at all.

Fokko commented on airflow/operators/python_operator.py:
```diff
@@ -330,13 +330,28 @@ def _write_string_args(self, filename):
+                    "Exception %s found while serializing argument
+                    object: %s on op_kwargs key %s ...skipping..."
+                    """ % (e, value, key)
+                    self.log.debug(msg)
```
I would raise this level. If you try to pass something, and it being dropped silently, that would be very confusing. What do you think? Any idea how often this happens?

Fokko commented on airflow/operators/python_operator.py:
```diff
@@ -330,13 +330,28 @@ def _write_string_args(self, filename):
+                    serializer.loads(serializer.dumps(value))
+                    if key not in invalid_args:
+                        kwargs[key] = value
+                except Exception as e:
```
`Exception` is a bit broad, can we be more specific here?

Fokko commented on airflow/operators/python_operator.py:
```diff
@@ -330,13 +330,28 @@ def _write_string_args(self, filename):
+        invalid_args = set(['dag', 'task', 'ti'])
         if self._pass_op_args():
+            kwargs = {}
+            for key, value in self.op_kwargs.items():
```
More from a code point of view, but I would do something like:
```python
MacBook-Pro-van-Fokko:home-analytics fokkodriesprong$ python2
Python 2.7.17 (default, Dec 23 2019, 21:25:33) 
[GCC 4.2.1 Compatible Apple LLVM 11.0.0 (clang-1100.0.33.16)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> invalid_args = {'dag', 'task', 'ti'}
>>> op_kwargs = {
...   'ti': 'Ignored',
...   'foo': 'bar'
... }
>>> 
>>> def serialize(self, item):
...     try:
...         return serializer.loads(serializer.dumps(value))
...     except Exception as e:
...         msg = """
...         "Exception %s found while serializing argument
...         object: %s on op_kwargs key %s ...skipping..."
...         """ % (e, value, key)
...         self.log.debug(msg)
... 
>>> # Replace it with something simple for the sake of showing how this works.
>>> def serialize(item):
...     return item
... 
>>> kwargs = {key: serialize(item) for key, item in op_kwargs.items() if key not in invalid_args}
>>> 
>>> kwargs
{'foo': 'bar'}
```
This makes it also easier to test the serialize function.

maganaluis replied:
@Fokko Thank you for taking time to review this PR, I've updated it based on your suggestions. 

```python
    def _write_args(self, input_filename):
        # serialize args to file
        if self.use_dill:
            serializer = dill
        else:
            serializer = pickle
        # some items from context can't be loaded in virtual env
        # see pr https://github.com/apache/airflow/pull/8256
        not_serializable = {'dag', 'task', 'ti', 'macros', 'task_instance', 'var'}
        if self._pass_op_args():
            kwargs = {key: value for key, value in self.op_kwargs.items()
                      if key not in not_serializable}
            with open(input_filename, 'wb') as f:
                arg_dict = ({'args': self.op_args, 'kwargs': kwargs})
                serializer.dump(arg_dict, f)
````

The invalid arguments here are SQLAlchemy db models, some fail to serialize, and some others (dag, task, and ti) fail to deserialize at run time due to reference to objects which are no longer present. See log.

[log_1.txt](https://github.com/apache/airflow/files/4572523/log_1.txt)

To solve this issues we would need to implement a __setstate__ and __getstate__ internal functions in those objects so they are properly serialized. This is not the scope of this PR and does represent considerable work, and tests. 

I've also added a pytest to ensure this gets tested in the future. 

```python
    def test_config_context(self):
        """
        This test ensures we can use dag_run from the context
        to access the configuration at run time that's being
        passed from the UI, CLI, and REST API.
        """
        self.dag.create_dagrun(
            run_id='manual__' + DEFAULT_DATE.isoformat(),
            execution_date=DEFAULT_DATE,
            start_date=DEFAULT_DATE,
            state=State.RUNNING,
            external_trigger=False,
        )

        def pass_function(**kwargs):
            kwargs['dag_run'].conf

        t = PythonVirtualenvOperator(task_id='config_dag_run', dag=self.dag,
                                     provide_context=True,
                                     python_callable=pass_function)
        t.run(start_date=DEFAULT_DATE, end_date=DEFAULT_DATE)
```

Fokko replied:
Nice, thanks for taking the time to sort this out

Fokko reviewed (commented):
It looks like a good fix. I would have expected to have more people would have encountered this issue. A couple of things:
- I don't think #6074 had been cherry-picked back to the 1.10 branch. 
- I've been a bit out of the loop, are we still using the `v1-10-test` branch?
- Maybe add a test? This will make sure that it won't happen in the future.

Fokko commented on tests/operators/test_python_operator.py:
```diff
@@ -340,6 +341,87 @@ def test_echo_env_variables(self):
+    def tearDown(self):
+        super(TestPythonVirtualenvOperator, self).tearDown()
+
+        session = Session()
```
```python
with create_session() as session:
        session.query(DagRun).delete()
        session.query(TI).delete()
```

Fokko commented on tests/operators/test_python_operator.py:
```diff
@@ -340,6 +341,87 @@ def test_echo_env_variables(self):
+
+        session.query(DagRun).delete()
+        session.query(TI).delete()
+        print(len(session.query(DagRun).all()))
```
Can you remove the print?

Fokko commented on tests/operators/test_python_operator.py:
```diff
@@ -340,6 +341,87 @@ def test_echo_env_variables(self):
+    def setUpClass(cls):
+        super(TestPythonVirtualenvOperator, cls).setUpClass()
+
+        session = Session()
```
```python
with create_session() as session:
        session.query(DagRun).delete()
        session.query(TI).delete()
```

Fokko reviewed (approved):
Small issues, apart from that; LGTM

ashb commented on tests/operators/test_python_operator.py:
```diff
@@ -340,6 +341,80 @@ def test_echo_env_variables(self):
+        with create_session() as session:
+            session.query(DagRun).delete()
+            session.query(TI).delete()
+            session.commit()
```
For consistency with other tests please use the helpers from https://github.com/apache/airflow/blob/v1-10-stable/tests/test_utils/db.py

ashb commented on tests/operators/test_python_operator.py:
```diff
@@ -340,6 +341,80 @@ def test_echo_env_variables(self):
+        t = PythonVirtualenvOperator(task_id='config_dag_run', dag=self.dag,
+                                     provide_context=True,
+                                     python_callable=pass_function)
+        t.run(start_date=DEFAULT_DATE, end_date=DEFAULT_DATE)
```
Does doing this create a _whole_ virtual environment? That takes quite a while doesn't it?

Could we perhaps just test the one function (`_write_args`) instead if that is the case

maganaluis replied:
This test runs at about 2 sec, that's not too slow...

maganaluis committed:
updated _write_args on PythonVirtualenvOperator

This change will ensure that provide_context=True works on v1.10.10 for PythonVirtualenvOperator
switched log message to debug on _write_args

switched log message to debug on _write_args
Clean up code and simplified

Clean up code and simplified the solution to ignore just the unserializable items
adding pytest for pythonvirtualenvoperator

this test ensures that we can use dag_run from context
Switch to test utils

Switched the session clear run and clear dags to test util functions.

Commit:
Fix task_instance_mutation_hook (#9910)

Fixes #9902

potiuk committed:
Fixes name of pre-commit cache for multiple branches (#10299)

(cherry-picked from commit 4aa05a7441803279ebae42572dcc89891323cd2a)

maganaluis committed:
Merge branch 'v1-10-stable' into v1-10-stable
